src/module/system/hotbar-macros.js

```diff
@@ -0,0 +1,284 @@
+import { ActorItemHelper, getChildItems } from "../actor/actor-inventory-utils.js";
+import { checkIcons } from "./enrichers/check.js";
+
+const createMacroFnLookup = {
+    Item: createItemMacro,
+    SkillCheck: createSkillCheckMacro,
+    SaveCheck: createSaveCheckMacro,
+    AbilityCheck: createAbilityCheckMacro
+};
+
+Hooks.on("hotbarDrop", async (bar, data, slot) => {
+    const createMacroFn = data && createMacroFnLookup[data.type];
+    if (createMacroFn) {
+        game.user.assignHotbarMacro(await createMacroFn(data), slot);
+    } else {
+        // silently ignore this hook, just in case someone else will pick it up.
+    }
+});
+
+/**
+ * Utility function for calling `Macro.create` only if an identical macro
+ * doesn't already exist. If one does, return that one instead.
+ *
+ * @param {Object} data Macro creation description.
+ * @returns {Promise<Macro>}
+ */
+function findElseCreateMacro(data) {
+    const existingMacro = game.macros.contents
+        .find(m => (m.name === data.name) && (m.command === data.command));
+
+    return existingMacro
+        ? Promise.resolve(existingMacro)
+        : Macro.create(data, { displaySheet: false });
+}
```

Comment:
Think we can just make this an async function and return `existingMacro` or call `Macro.create`, without manually creating promises

Reply:
Yep. On second look, this can be greatly simplified.